subprojects/architecture-test/src/test/java/org/gradle/architecture/test/InternalNullabilityTest.java

```diff
@@ -0,0 +1,664 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.architecture.test;
+
+import com.tngtech.archunit.base.DescribedPredicate;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchRule;
+import org.gradle.api.NonNullApi;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.tngtech.archunit.base.DescribedPredicate.not;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static org.gradle.architecture.test.ArchUnitFixture.beAnnotatedOrInPackageAnnotatedWith;
+import static org.gradle.architecture.test.ArchUnitFixture.classes_not_written_in_kotlin;
+import static org.gradle.architecture.test.ArchUnitFixture.gradleInternalApi;
+
+@AnalyzeClasses(packages = "org.gradle")
+public class InternalNullabilityTest {
+
+    /**
+     * List of packages that are not yet annotated with {@link NonNullApi}.
+     * This list should never grow and should be empty at some point.
+     */
+    private static final Set<String> PENDING_ANNOTATION = new HashSet<>(Arrays.asList(
+        "org.gradle.api.distribution.internal",
```

Comment:
Could we use the regular arch unit freezing of rules instead of maintaining this list? That way we can ensure that newly added classes are annotated properly and add some more pressure for moving in the right direction.